src/openslide-vendor-philips-isyntax.c

```diff
@@ -0,0 +1,397 @@
+            // will need to have allocator per size. Alternatively, implement allocator freeing after
+            // all tiles have been freed, and track isyntax_t per tile so we can access allocator.
+            philips_isyntax_global_cache_ptr = malloc(sizeof(*philips_isyntax_global_cache_ptr));
+            ASSERT_OK(libisyntax_cache_create("global_cache_list", cache_size, &philips_isyntax_global_cache_ptr->cache));
+        }
+        data->cache = philips_isyntax_global_cache_ptr;
+        g_mutex_unlock(&static_open_mutex);
+    } else {
+        ASSERT_OK(libisyntax_cache_create("cache_list", cache_size, &data->cache->cache));
+    }
+    // Link the cache (local or global) to the isyntax file.
+    libisyntax_cache_inject(data->cache->cache, data->isyntax);
+
+    LOG_VAR("%d", libisyntax_get_is_mpp_known(data->isyntax));
+    if (libisyntax_get_is_mpp_known(data->isyntax)) {
+        double mpp_x = libisyntax_get_mpp_x(data->isyntax);
+        double mpp_y = libisyntax_get_mpp_y(data->isyntax);
+        LOG_VAR("%f", mpp_x);
+        LOG_VAR("%f", mpp_y);
+        add_float_property(osr, OPENSLIDE_PROPERTY_NAME_MPP_X, mpp_x);
+        add_float_property(osr, OPENSLIDE_PROPERTY_NAME_MPP_Y, mpp_y);
+        const float float_equals_tolerance = 1e-5;
+        if (fabs(mpp_x - mpp_y) < float_equals_tolerance) {
+            // Compute objective power from microns-per-pixel, see e.g. table in "Scan Performance" here:
+            // https://www.microscopesinternational.com/blog/20170928-whichobjective.aspx
+            float objective_power = 10.0f / mpp_x;
+            LOG_VAR("%f", objective_power);
+            add_float_property(osr, OPENSLIDE_PROPERTY_NAME_OBJECTIVE_POWER, objective_power);
+        }
+    }
+
+    // Find wsi image. Extracting other images not supported. Assuming only one wsi.
+    int wsi_image_idx = libisyntax_get_wsi_image_index(data->isyntax);
+    LOG_VAR("%d", wsi_image_idx);
+    const isyntax_image_t* wsi_image = libisyntax_get_image(data->isyntax, wsi_image_idx);
+
+    // Store openslide information about each level.
+    osr->level_count = libisyntax_image_get_level_count(wsi_image);
+    osr->levels = malloc(sizeof(philips_isyntax_level*) * osr->level_count);
+    for (int i = 0; i < osr->level_count; ++i) {
+        philips_isyntax_level* level = malloc(sizeof(philips_isyntax_level));
+        level->isyntax_level = libisyntax_image_get_level(wsi_image, i);
+        level->base.downsample = libisyntax_level_get_downsample_factor(level->isyntax_level);
+        level->base.tile_w = libisyntax_get_tile_width(data->isyntax);
+        level->base.tile_h = libisyntax_get_tile_height(data->isyntax);
+        level->base.w = libisyntax_level_get_width_in_tiles(level->isyntax_level) * level->base.tile_w;
+        level->base.h = libisyntax_level_get_height_in_tiles(level->isyntax_level) * level->base.tile_h;
```

Comment:
Well, it's somewhat unclear to me (even now) what the actual 'correct' dimensions of the image should be for each individual level, because the information from various sources (including official documentation, header fields) are somewhat confusing to interpret.

Note that the grid 'shifts' across the levels, meaning that the parent tile is never a perfect 'downsample' of its children. 

There is a field in the XML header containing the full image dimensions for level 0, which includes the padding that has been added around the images in the encoding process. This number might not align fully with the grid size. Also see here: 
https://github.com/amspath/libisyntax/issues/20
https://github.com/amspath/libisyntax/pull/21

It's true however that all the grid tiles themselves as encoded in the file are 'complete'. So in that sense, there is no chance of a partial last column/row.